{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ada3dd6c-e196-4236-87e2-59bd12ee93c9",
   "metadata": {},
   "source": [
    "# Horcrux Joystick 입력 학습 진행"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3d86890-4e2f-4c3e-a30c-b281c5de415b",
   "metadata": {},
   "source": [
    "## 필요 패키지 import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ca9f4abf-b333-49c3-9ea6-22f049e2eb51",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import gymnasium as gym\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# 조이스틱 환경 삽입\n",
    "from horcrux_terrain_v1.envs import PlaneJoyWorld\n",
    "from horcrux_terrain_v1.envs import PlaneWorld\n",
    "\n",
    "# Ray 패키지 삽입\n",
    "import ray\n",
    "from ray.rllib.algorithms.algorithm import Algorithm\n",
    "from ray.rllib.algorithms.sac import SACConfig\n",
    "\n",
    "from ray.tune.registry import register_env"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6517ec57-16da-4789-a76f-2c77998e7a5e",
   "metadata": {},
   "source": [
    "## Ray 실행"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df31cbbe-264f-4298-a2f1-471cf823d906",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-03-13 15:21:11,550\tINFO worker.py:1777 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0e4d6b1151c1487dbf38b154042ba1f8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<div class=\"lm-Widget p-Widget lm-Panel p-Panel jp-Cell-outputWrapper\">\n",
       "    <div style=\"margin-left: 50px;display: flex;flex-direction: row;align-items: center\">\n",
       "        <div class=\"jp-RenderedHTMLCommon\" style=\"display: flex; flex-direction: row;\">\n",
       "  <svg viewBox=\"0 0 567 224\" fill=\"none\" xmlns=\"http://www.w3.org/2000/svg\" style=\"height: 3em;\">\n",
       "    <g clip-path=\"url(#clip0_4338_178347)\">\n",
       "        <path d=\"M341.29 165.561H355.29L330.13 129.051C345.63 123.991 354.21 112.051 354.21 94.2307C354.21 71.3707 338.72 58.1807 311.88 58.1807H271V165.561H283.27V131.661H311.8C314.25 131.661 316.71 131.501 319.01 131.351L341.25 165.561H341.29ZM283.29 119.851V70.0007H311.82C331.3 70.0007 342.34 78.2907 342.34 94.5507C342.34 111.271 331.34 119.861 311.82 119.861L283.29 119.851ZM451.4 138.411L463.4 165.561H476.74L428.74 58.1807H416L367.83 165.561H380.83L392.83 138.411H451.4ZM446.19 126.601H398L422 72.1407L446.24 126.601H446.19ZM526.11 128.741L566.91 58.1807H554.35L519.99 114.181L485.17 58.1807H472.44L514.01 129.181V165.541H526.13V128.741H526.11Z\" fill=\"var(--jp-ui-font-color0)\"/>\n",
       "        <path d=\"M82.35 104.44C84.0187 97.8827 87.8248 92.0678 93.1671 87.9146C98.5094 83.7614 105.083 81.5067 111.85 81.5067C118.617 81.5067 125.191 83.7614 130.533 87.9146C135.875 92.0678 139.681 97.8827 141.35 104.44H163.75C164.476 101.562 165.622 98.8057 167.15 96.2605L127.45 56.5605C121.071 60.3522 113.526 61.6823 106.235 60.3005C98.9443 58.9187 92.4094 54.9203 87.8602 49.0574C83.3109 43.1946 81.0609 35.8714 81.5332 28.4656C82.0056 21.0599 85.1679 14.0819 90.4252 8.8446C95.6824 3.60726 102.672 0.471508 110.08 0.0272655C117.487 -0.416977 124.802 1.86091 130.647 6.4324C136.493 11.0039 140.467 17.5539 141.821 24.8501C143.175 32.1463 141.816 39.6859 138 46.0505L177.69 85.7505C182.31 82.9877 187.58 81.4995 192.962 81.4375C198.345 81.3755 203.648 82.742 208.33 85.3976C213.012 88.0532 216.907 91.9029 219.616 96.5544C222.326 101.206 223.753 106.492 223.753 111.875C223.753 117.258 222.326 122.545 219.616 127.197C216.907 131.848 213.012 135.698 208.33 138.353C203.648 141.009 198.345 142.375 192.962 142.313C187.58 142.251 182.31 140.763 177.69 138L138 177.7C141.808 184.071 143.155 191.614 141.79 198.91C140.424 206.205 136.44 212.75 130.585 217.313C124.731 221.875 117.412 224.141 110.004 223.683C102.596 223.226 95.6103 220.077 90.3621 214.828C85.1139 209.58 81.9647 202.595 81.5072 195.187C81.0497 187.779 83.3154 180.459 87.878 174.605C92.4405 168.751 98.9853 164.766 106.281 163.401C113.576 162.035 121.119 163.383 127.49 167.19L167.19 127.49C165.664 124.941 164.518 122.182 163.79 119.3H141.39C139.721 125.858 135.915 131.673 130.573 135.826C125.231 139.98 118.657 142.234 111.89 142.234C105.123 142.234 98.5494 139.98 93.2071 135.826C87.8648 131.673 84.0587 125.858 82.39 119.3H60C58.1878 126.495 53.8086 132.78 47.6863 136.971C41.5641 141.163 34.1211 142.972 26.7579 142.059C19.3947 141.146 12.6191 137.574 7.70605 132.014C2.79302 126.454 0.0813599 119.29 0.0813599 111.87C0.0813599 104.451 2.79302 97.2871 7.70605 91.7272C12.6191 86.1673 19.3947 82.5947 26.7579 81.6817C34.1211 80.7686 41.5641 82.5781 47.6863 86.7696C53.8086 90.9611 58.1878 97.2456 60 104.44H82.35ZM100.86 204.32C103.407 206.868 106.759 208.453 110.345 208.806C113.93 209.159 117.527 208.258 120.522 206.256C123.517 204.254 125.725 201.276 126.771 197.828C127.816 194.38 127.633 190.677 126.253 187.349C124.874 184.021 122.383 181.274 119.205 179.577C116.027 177.88 112.359 177.337 108.826 178.042C105.293 178.746 102.113 180.654 99.8291 183.44C97.5451 186.226 96.2979 189.718 96.3 193.32C96.2985 195.364 96.7006 197.388 97.4831 199.275C98.2656 201.163 99.4132 202.877 100.86 204.32ZM204.32 122.88C206.868 120.333 208.453 116.981 208.806 113.396C209.159 109.811 208.258 106.214 206.256 103.219C204.254 100.223 201.275 98.0151 197.827 96.97C194.38 95.9249 190.676 96.1077 187.348 97.4873C184.02 98.8669 181.274 101.358 179.577 104.536C177.879 107.714 177.337 111.382 178.041 114.915C178.746 118.448 180.653 121.627 183.439 123.911C186.226 126.195 189.717 127.443 193.32 127.44C195.364 127.443 197.388 127.042 199.275 126.259C201.163 125.476 202.878 124.328 204.32 122.88ZM122.88 19.4205C120.333 16.8729 116.981 15.2876 113.395 14.9347C109.81 14.5817 106.213 15.483 103.218 17.4849C100.223 19.4868 98.0146 22.4654 96.9696 25.9131C95.9245 29.3608 96.1073 33.0642 97.4869 36.3922C98.8665 39.7202 101.358 42.4668 104.535 44.1639C107.713 45.861 111.381 46.4036 114.914 45.6992C118.447 44.9949 121.627 43.0871 123.911 40.301C126.195 37.515 127.442 34.0231 127.44 30.4205C127.44 28.3772 127.038 26.3539 126.255 24.4664C125.473 22.5788 124.326 20.8642 122.88 19.4205ZM19.42 100.86C16.8725 103.408 15.2872 106.76 14.9342 110.345C14.5813 113.93 15.4826 117.527 17.4844 120.522C19.4863 123.518 22.4649 125.726 25.9127 126.771C29.3604 127.816 33.0638 127.633 36.3918 126.254C39.7198 124.874 42.4664 122.383 44.1635 119.205C45.8606 116.027 46.4032 112.359 45.6988 108.826C44.9944 105.293 43.0866 102.114 40.3006 99.8296C37.5145 97.5455 34.0227 96.2983 30.42 96.3005C26.2938 96.3018 22.337 97.9421 19.42 100.86ZM100.86 100.86C98.3125 103.408 96.7272 106.76 96.3742 110.345C96.0213 113.93 96.9226 117.527 98.9244 120.522C100.926 123.518 103.905 125.726 107.353 126.771C110.8 127.816 114.504 127.633 117.832 126.254C121.16 124.874 123.906 122.383 125.604 119.205C127.301 116.027 127.843 112.359 127.139 108.826C126.434 105.293 124.527 102.114 121.741 99.8296C118.955 97.5455 115.463 96.2983 111.86 96.3005C109.817 96.299 107.793 96.701 105.905 97.4835C104.018 98.2661 102.303 99.4136 100.86 100.86Z\" fill=\"#00AEEF\"/>\n",
       "    </g>\n",
       "    <defs>\n",
       "        <clipPath id=\"clip0_4338_178347\">\n",
       "            <rect width=\"566.93\" height=\"223.75\" fill=\"white\"/>\n",
       "        </clipPath>\n",
       "    </defs>\n",
       "  </svg>\n",
       "</div>\n",
       "\n",
       "        <table class=\"jp-RenderedHTMLCommon\" style=\"border-collapse: collapse;color: var(--jp-ui-font-color1);font-size: var(--jp-ui-font-size1);\">\n",
       "    <tr>\n",
       "        <td style=\"text-align: left\"><b>Python version:</b></td>\n",
       "        <td style=\"text-align: left\"><b>3.11.11</b></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td style=\"text-align: left\"><b>Ray version:</b></td>\n",
       "        <td style=\"text-align: left\"><b>2.36.1</b></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "    <td style=\"text-align: left\"><b>Dashboard:</b></td>\n",
       "    <td style=\"text-align: left\"><b><a href=\"http://127.0.0.1:8265\" target=\"_blank\">http://127.0.0.1:8265</a></b></td>\n",
       "</tr>\n",
       "\n",
       "</table>\n",
       "\n",
       "    </div>\n",
       "</div>\n"
      ],
      "text/plain": [
       "RayContext(dashboard_url='127.0.0.1:8265', python_version='3.11.11', ray_version='2.36.1', ray_commit='999f7668b65dbd208817b3671b441fd97fca755c')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m Stack (most recent call first):\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376 in __init__\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183 in _load_all_bundled_plugins\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190 in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 241 in _call_with_frames_removed\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap_external>\", line 940 in exec_module\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 690 in _load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1147 in _find_and_load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1176 in _find_and_load\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19 in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 241 in _call_with_frames_removed\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap_external>\", line 940 in exec_module\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 690 in _load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1147 in _find_and_load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1176 in _find_and_load\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1 in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 241 in _call_with_frames_removed\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap_external>\", line 940 in exec_module\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 690 in _load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1147 in _find_and_load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1176 in _find_and_load\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 241 in _call_with_frames_removed\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1126 in _find_and_load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1176 in _find_and_load\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2 in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 241 in _call_with_frames_removed\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap_external>\", line 940 in exec_module\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 690 in _load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1147 in _find_and_load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1176 in _find_and_load\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3 in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 241 in _call_with_frames_removed\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap_external>\", line 940 in exec_module\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 690 in _load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1147 in _find_and_load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1176 in _find_and_load\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 241 in _call_with_frames_removed\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1126 in _find_and_load_unlocked\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"<frozen importlib._bootstrap>\", line 1176 in _find_and_load\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225 in _deserialize_pickle5_data\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235 in _deserialize_msgpack_data\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280 in _deserialize_object\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423 in deserialize_objects\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\worker.py\", line 811 in deserialize_objects\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\worker.py\", line 887 in main_loop\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\workers\\default_worker.py\", line 289 in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     obj = self._deserialize_object(data, metadata, object_ref)\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     obj = pickle.loads(in_band)\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m           ^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     import mujoco\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     _load_all_bundled_plugins()\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m     self._handle = _dlopen(self._name, mode)\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m                    ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m Exception raised in creation task: The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=39896, ip=127.0.0.1, actor_id=a68a48717d09dbadc5fa501401000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x000001A413DB31D0>)\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m   File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "\u001b[36m(RolloutWorker pid=39896)\u001b[0m traceback: Traceback (most recent call last):\n"
     ]
    }
   ],
   "source": [
    "# import socket\n",
    "# import psutil\n",
    "\n",
    "# conn_ip = \"\"\n",
    "# interfaces = psutil.net_if_addrs()\n",
    "# for interface_name, addresses in interfaces.items():\n",
    "#     if \"openvpn\" in interface_name.lower() and \"tap\" in interface_name.lower():\n",
    "#         snicaddrs = interfaces[str(interface_name)]\n",
    "#         for addrfamily in snicaddrs:\n",
    "#             if addrfamily.family == socket.AF_INET:\n",
    "#                 conn_ip = addrfamily.address\n",
    "\n",
    "# print(f\"Connection IP: {conn_ip}\")\n",
    "# # 해당 init을 통해서 VPN을 통한 외부 접속 가능함.\n",
    "# ray.init(dashboard_host=conn_ip, dashboard_port=8265)\n",
    "\n",
    "ray.init()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39407944-23a9-42b0-854b-c19f1c43bcdc",
   "metadata": {},
   "source": [
    "## Gym 환경 등록하기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6809e905-5daf-45c4-919f-98e53d97572e",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_config = {\n",
    "    \"forward_reward_weight\": 6.5,\n",
    "    \"side_cost_weight\": 2.0,\n",
    "    \"unhealthy_max_steps\": 100,\n",
    "    \"healthy_reward\": 0.5,\n",
    "    \"healthy_roll_range\": (-35,35),\n",
    "    \"terminating_roll_range\": (-85,85),\n",
    "    \"rotation_norm_cost_weight\": 0.01,\n",
    "    \"rotation_orientation_cost_weight\": 1.2,\n",
    "    \"termination_reward\": 0,\n",
    "    \"gait_params\": (30, 30, 60, 60, 0),\n",
    "    \"use_friction_chg\": True,\n",
    "    \"joy_input_random\": True,\n",
    "}\n",
    "\n",
    "# JoyWorld\n",
    "register_env(\"joy-v1\", lambda config: PlaneJoyWorld( forward_reward_weight=env_config[\"forward_reward_weight\"], \n",
    "                                                     side_cost_weight=env_config[\"side_cost_weight\"], \n",
    "                                                     unhealthy_max_steps=env_config[\"unhealthy_max_steps\"],\n",
    "                                                     healthy_reward=env_config[\"healthy_reward\"], \n",
    "                                                     healthy_roll_range=env_config[\"healthy_roll_range\"],\n",
    "                                                     terminating_roll_range=env_config[\"terminating_roll_range\"],\n",
    "                                                     rotation_norm_cost_weight=env_config[\"rotation_norm_cost_weight\"],\n",
    "                                                     rotation_orientation_cost_weight=env_config[\"rotation_orientation_cost_weight\"],\n",
    "                                                     termination_reward=env_config[\"termination_reward\"],\n",
    "                                                     gait_params=env_config[\"gait_params\"],\n",
    "                                                     use_friction_chg=env_config[\"use_friction_chg\"],\n",
    "                                                     joy_input_random=env_config[\"joy_input_random\"],\n",
    "                                                   )\n",
    "            )\n",
    "\n",
    "# Plane\n",
    "register_env(\"plane-v1\", lambda config: PlaneWorld(forward_reward_weight=env_config[\"forward_reward_weight\"], \n",
    "                                                 side_cost_weight=env_config[\"side_cost_weight\"], \n",
    "                                                 unhealthy_max_steps=env_config[\"unhealthy_max_steps\"], \n",
    "                                                 healthy_reward=env_config[\"healthy_reward\"],\n",
    "                                                 healthy_roll_range=env_config[\"healthy_roll_range\"],\n",
    "                                                 terminating_roll_range=env_config[\"terminating_roll_range\"],\n",
    "                                                 rotation_norm_cost_weight=env_config[\"rotation_norm_cost_weight\"],\n",
    "                                                 rotation_orientation_cost_weight=env_config[\"rotation_orientation_cost_weight\"],\n",
    "                                                 termination_reward=env_config[\"termination_reward\"],\n",
    "                                                 use_friction_chg=env_config[\"use_friction_chg\"],\n",
    "                                                 gait_params=env_config[\"gait_params\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b260cf6f-c922-49e2-96ff-6a94536a1bd0",
   "metadata": {},
   "source": [
    "## 학습된 알고리즘 불러오기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d211abe2-7e3f-4825-abc9-521ca36200b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# algo = Algorithm.from_checkpoint(\"./Paper_agents/good/Linear/SAC_layer_512_5_32_Linear_restart_final\")\n",
    "\n",
    "# trained_weights = algo.get_weights()\n",
    "# algo.cleanup()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fbf2134-3167-46f1-b28b-7d305065f559",
   "metadata": {},
   "source": [
    "## 학습 알고리즘 설정하기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "436edebc-bcaa-45fa-9941-58ad2868655b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\algorithms\\algorithm.py:555: RayDeprecationWarning: This API is deprecated and may be removed in future Ray releases. You could suppress this warning by setting env variable PYTHONWARNINGS=\"ignore::DeprecationWarning\"\n",
      "`UnifiedLogger` will be removed in Ray 2.7.\n",
      "  return UnifiedLogger(config, logdir, loggers=None)\n",
      "c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\tune\\logger\\unified.py:53: RayDeprecationWarning: This API is deprecated and may be removed in future Ray releases. You could suppress this warning by setting env variable PYTHONWARNINGS=\"ignore::DeprecationWarning\"\n",
      "The `JsonLogger interface is deprecated in favor of the `ray.tune.json.JsonLoggerCallback` interface and will be removed in Ray 2.7.\n",
      "  self._loggers.append(cls(self.config, self.logdir, self.trial))\n",
      "c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\tune\\logger\\unified.py:53: RayDeprecationWarning: This API is deprecated and may be removed in future Ray releases. You could suppress this warning by setting env variable PYTHONWARNINGS=\"ignore::DeprecationWarning\"\n",
      "The `CSVLogger interface is deprecated in favor of the `ray.tune.csv.CSVLoggerCallback` interface and will be removed in Ray 2.7.\n",
      "  self._loggers.append(cls(self.config, self.logdir, self.trial))\n",
      "c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\tune\\logger\\unified.py:53: RayDeprecationWarning: This API is deprecated and may be removed in future Ray releases. You could suppress this warning by setting env variable PYTHONWARNINGS=\"ignore::DeprecationWarning\"\n",
      "The `TBXLogger interface is deprecated in favor of the `ray.tune.tensorboardx.TBXLoggerCallback` interface and will be removed in Ray 2.7.\n",
      "  self._loggers.append(cls(self.config, self.logdir, self.trial))\n",
      "2025-03-13 15:21:25,387\tERROR actor_manager.py:523 -- Ray error, taking actor 1 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=36256, ip=127.0.0.1, actor_id=eadd0723e75c6706e190a27d01000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000013B7FFCB610>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,395\tERROR actor_manager.py:523 -- Ray error, taking actor 2 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=32748, ip=127.0.0.1, actor_id=a1bd393d71b3bfef8c36958601000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000020A88BBB610>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,397\tERROR actor_manager.py:523 -- Ray error, taking actor 3 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=37328, ip=127.0.0.1, actor_id=4eb21897a38822c4a44c7fe301000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000024E6E5FB610>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,397\tERROR actor_manager.py:523 -- Ray error, taking actor 4 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=39364, ip=127.0.0.1, actor_id=8e5c167382a95fd23db98f5c01000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000019F15D34050>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,399\tERROR actor_manager.py:523 -- Ray error, taking actor 5 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=33876, ip=127.0.0.1, actor_id=323e4b35305f12df6e9e3e7701000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000015DEA07D010>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,399\tERROR actor_manager.py:523 -- Ray error, taking actor 6 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=35076, ip=127.0.0.1, actor_id=64e8db690238f1c128135efd01000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000017E5787B610>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,399\tERROR actor_manager.py:523 -- Ray error, taking actor 7 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=17396, ip=127.0.0.1, actor_id=f1ae8e7945faa00ac2203a2401000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x000001738D5CB610>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,399\tERROR actor_manager.py:523 -- Ray error, taking actor 8 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=35736, ip=127.0.0.1, actor_id=0837e440792dc65847fb1c9001000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000023E068A8B90>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,399\tERROR actor_manager.py:523 -- Ray error, taking actor 9 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=13504, ip=127.0.0.1, actor_id=4f6f3839fa32a390b430017501000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x000001D9D98E4050>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,399\tERROR actor_manager.py:523 -- Ray error, taking actor 10 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=36168, ip=127.0.0.1, actor_id=56fbd4548f62ffdc1842379a01000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000017B41185850>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,399\tERROR actor_manager.py:523 -- Ray error, taking actor 11 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=39896, ip=127.0.0.1, actor_id=a68a48717d09dbadc5fa501401000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x000001A413DB31D0>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,399\tERROR actor_manager.py:523 -- Ray error, taking actor 12 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=33864, ip=127.0.0.1, actor_id=3614a10d1295938c0893797301000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000028E38DD4050>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,408\tERROR actor_manager.py:523 -- Ray error, taking actor 13 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=30504, ip=127.0.0.1, actor_id=be0ac407a7d6da9572f1852a01000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000017BB4084050>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,408\tERROR actor_manager.py:523 -- Ray error, taking actor 14 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=35568, ip=127.0.0.1, actor_id=ce361ed91d494e9c14cd21ee01000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x00000173DCEEB610>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,408\tERROR actor_manager.py:523 -- Ray error, taking actor 15 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=35520, ip=127.0.0.1, actor_id=f25fa1ba77b174342376225d01000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x000001AD4D987210>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "2025-03-13 15:21:25,408\tERROR actor_manager.py:523 -- Ray error, taking actor 16 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=28836, ip=127.0.0.1, actor_id=e76ff3525db38176f6728b1701000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000017BFDF74050>)\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n",
      "  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\n",
      "ray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
      "traceback: Traceback (most recent call last):\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n",
      "    obj = self._deserialize_object(data, metadata, object_ref)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n",
      "    return self._deserialize_msgpack_data(data, metadata_fields)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n",
      "    python_objects = self._deserialize_pickle5_data(pickle5_data)\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n",
      "    obj = pickle.loads(in_band)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n",
      "    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n",
      "  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n",
      "    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n",
      "    import mujoco\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n",
      "    _load_all_bundled_plugins()\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n",
      "    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n",
      "    self._handle = _dlopen(self._name, mode)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "OSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n"
     ]
    },
    {
     "ename": "RaySystemError",
     "evalue": "System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\ntraceback: Traceback (most recent call last):\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n    obj = self._deserialize_object(data, metadata, object_ref)\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n    return self._deserialize_msgpack_data(data, metadata_fields)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n    python_objects = self._deserialize_pickle5_data(pickle5_data)\n                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n    obj = pickle.loads(in_band)\n          ^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n    import mujoco\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n    _load_all_bundled_plugins()\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n    self._handle = _dlopen(self._name, mode)\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^\nOSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mActorDiedError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\env\\env_runner_group.py:189\u001b[39m, in \u001b[36mEnvRunnerGroup.__init__\u001b[39m\u001b[34m(self, env_creator, validate_env, default_policy_class, config, num_env_runners, local_env_runner, logdir, _setup, tune_trial_id, num_workers, local_worker)\u001b[39m\n\u001b[32m    188\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m189\u001b[39m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_setup\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    190\u001b[39m \u001b[43m        \u001b[49m\u001b[43mvalidate_env\u001b[49m\u001b[43m=\u001b[49m\u001b[43mvalidate_env\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    191\u001b[39m \u001b[43m        \u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m=\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    192\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnum_env_runners\u001b[49m\u001b[43m=\u001b[49m\u001b[43mnum_env_runners\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    193\u001b[39m \u001b[43m        \u001b[49m\u001b[43mlocal_env_runner\u001b[49m\u001b[43m=\u001b[49m\u001b[43mlocal_env_runner\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    194\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    195\u001b[39m \u001b[38;5;66;03m# EnvRunnerGroup creation possibly fails, if some (remote) workers cannot\u001b[39;00m\n\u001b[32m    196\u001b[39m \u001b[38;5;66;03m# be initialized properly (due to some errors in the EnvRunners's\u001b[39;00m\n\u001b[32m    197\u001b[39m \u001b[38;5;66;03m# constructor).\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\env\\env_runner_group.py:259\u001b[39m, in \u001b[36mEnvRunnerGroup._setup\u001b[39m\u001b[34m(self, validate_env, config, num_env_runners, local_env_runner)\u001b[39m\n\u001b[32m    258\u001b[39m \u001b[38;5;66;03m# Create a number of @ray.remote workers.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m259\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43madd_workers\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    260\u001b[39m \u001b[43m    \u001b[49m\u001b[43mnum_env_runners\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    261\u001b[39m \u001b[43m    \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[43m=\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43mvalidate_env_runners_after_construction\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    262\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    264\u001b[39m \u001b[38;5;66;03m# If num_env_runners > 0 and we don't have an env on the local worker,\u001b[39;00m\n\u001b[32m    265\u001b[39m \u001b[38;5;66;03m# get the observation- and action spaces for each policy from\u001b[39;00m\n\u001b[32m    266\u001b[39m \u001b[38;5;66;03m# the first remote worker (which does have an env).\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\env\\env_runner_group.py:792\u001b[39m, in \u001b[36mEnvRunnerGroup.add_workers\u001b[39m\u001b[34m(self, num_workers, validate)\u001b[39m\n\u001b[32m    791\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m result.ok:\n\u001b[32m--> \u001b[39m\u001b[32m792\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m result.get()\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\utils\\actor_manager.py:500\u001b[39m, in \u001b[36mFaultTolerantActorManager._fetch_result\u001b[39m\u001b[34m(self, remote_actor_ids, remote_calls, tags, timeout_seconds, return_obj_refs, mark_healthy)\u001b[39m\n\u001b[32m    499\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m500\u001b[39m     result = \u001b[43mray\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mready\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    501\u001b[39m     remote_results.add_result(actor_id, ResultOrError(result=result), tag)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\auto_init_hook.py:21\u001b[39m, in \u001b[36mwrap_auto_init.<locals>.auto_init_wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m     20\u001b[39m auto_init_ray()\n\u001b[32m---> \u001b[39m\u001b[32m21\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\client_mode_hook.py:103\u001b[39m, in \u001b[36mclient_mode_hook.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    102\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(ray, func.\u001b[34m__name__\u001b[39m)(*args, **kwargs)\n\u001b[32m--> \u001b[39m\u001b[32m103\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\worker.py:2664\u001b[39m, in \u001b[36mget\u001b[39m\u001b[34m(object_refs, timeout)\u001b[39m\n\u001b[32m   2663\u001b[39m \u001b[38;5;66;03m# TODO(ujvl): Consider how to allow user to retrieve the ready objects.\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m2664\u001b[39m values, debugger_breakpoint = \u001b[43mworker\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_objects\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobject_refs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   2665\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m i, value \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(values):\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\worker.py:873\u001b[39m, in \u001b[36mWorker.get_objects\u001b[39m\u001b[34m(self, object_refs, timeout, return_exceptions)\u001b[39m\n\u001b[32m    872\u001b[39m             \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m                 \u001b[38;5;28;01mraise\u001b[39;00m value\n\u001b[32m    875\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m values, debugger_breakpoint\n",
      "\u001b[31mActorDiedError\u001b[39m: The actor died because of an error raised in its creation task, \u001b[36mray::RolloutWorker.__init__()\u001b[39m (pid=36256, ip=127.0.0.1, actor_id=eadd0723e75c6706e190a27d01000000, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x0000013B7FFCB610>)\n  File \"python\\ray\\_raylet.pyx\", line 1806, in ray._raylet.execute_task\n  File \"python\\ray\\_raylet.pyx\", line 1840, in ray._raylet.execute_task\n  File \"python\\ray\\_raylet.pyx\", line 943, in ray._raylet.raise_if_dependency_failed\nray.exceptions.RaySystemError: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\ntraceback: Traceback (most recent call last):\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n    obj = self._deserialize_object(data, metadata, object_ref)\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n    return self._deserialize_msgpack_data(data, metadata_fields)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n    python_objects = self._deserialize_pickle5_data(pickle5_data)\n                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n    obj = pickle.loads(in_band)\n          ^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n    import mujoco\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n    _load_all_bundled_plugins()\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n    self._handle = _dlopen(self._name, mode)\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^\nOSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[31mRaySystemError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[4]\u001b[39m\u001b[32m, line 51\u001b[39m\n\u001b[32m     15\u001b[39m config.env_runners(num_env_runners = total_workers, num_cpus_per_env_runner = \u001b[32m1\u001b[39m, num_gpus_per_env_runner = \u001b[32m0.3\u001b[39m/(total_workers))\n\u001b[32m     16\u001b[39m config.training(\n\u001b[32m     17\u001b[39m     gamma=\u001b[32m0.95\u001b[39m,\n\u001b[32m     18\u001b[39m     replay_buffer_config={\n\u001b[32m   (...)\u001b[39m\u001b[32m     48\u001b[39m     num_steps_sampled_before_learning_starts = \u001b[32m4096\u001b[39m,\n\u001b[32m     49\u001b[39m )\n\u001b[32m---> \u001b[39m\u001b[32m51\u001b[39m algo = \u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43mbuild\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\algorithms\\algorithm_config.py:889\u001b[39m, in \u001b[36mAlgorithmConfig.build\u001b[39m\u001b[34m(self, env, logger_creator, use_copy)\u001b[39m\n\u001b[32m    886\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m.algo_class, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m    887\u001b[39m     algo_class = get_trainable_cls(\u001b[38;5;28mself\u001b[39m.algo_class)\n\u001b[32m--> \u001b[39m\u001b[32m889\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43malgo_class\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    890\u001b[39m \u001b[43m    \u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43muse_copy\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    891\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlogger_creator\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mlogger_creator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    892\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\algorithms\\sac\\sac.py:537\u001b[39m, in \u001b[36mSAC.__init__\u001b[39m\u001b[34m(self, *args, **kwargs)\u001b[39m\n\u001b[32m    535\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m__init__\u001b[39m(\u001b[38;5;28mself\u001b[39m, *args, **kwargs):\n\u001b[32m    536\u001b[39m     \u001b[38;5;28mself\u001b[39m._allow_unknown_subkeys += [\u001b[33m\"\u001b[39m\u001b[33mpolicy_model_config\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mq_model_config\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m--> \u001b[39m\u001b[32m537\u001b[39m     \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[34;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\algorithms\\algorithm.py:571\u001b[39m, in \u001b[36mAlgorithm.__init__\u001b[39m\u001b[34m(self, config, env, logger_creator, **kwargs)\u001b[39m\n\u001b[32m    568\u001b[39m \u001b[38;5;66;03m# Evaluation EnvRunnerGroup and metrics last returned by `self.evaluate()`.\u001b[39;00m\n\u001b[32m    569\u001b[39m \u001b[38;5;28mself\u001b[39m.eval_env_runner_group: Optional[EnvRunnerGroup] = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m571\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[34;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[32m    572\u001b[39m \u001b[43m    \u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m=\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    573\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlogger_creator\u001b[49m\u001b[43m=\u001b[49m\u001b[43mlogger_creator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    574\u001b[39m \u001b[43m    \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    575\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\tune\\trainable\\trainable.py:158\u001b[39m, in \u001b[36mTrainable.__init__\u001b[39m\u001b[34m(self, config, logger_creator, storage)\u001b[39m\n\u001b[32m    154\u001b[39m     logger.debug(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mStorageContext on the TRAINABLE:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mstorage\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m    156\u001b[39m \u001b[38;5;28mself\u001b[39m._open_logfiles(stdout_file, stderr_file)\n\u001b[32m--> \u001b[39m\u001b[32m158\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43msetup\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    159\u001b[39m setup_time = time.time() - \u001b[38;5;28mself\u001b[39m._start_time\n\u001b[32m    160\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m setup_time > SETUP_TIME_THRESHOLD:\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\algorithms\\algorithm.py:657\u001b[39m, in \u001b[36mAlgorithm.setup\u001b[39m\u001b[34m(self, config)\u001b[39m\n\u001b[32m    654\u001b[39m     \u001b[38;5;28mself\u001b[39m.config.off_policy_estimation_methods = ope_dict\n\u001b[32m    656\u001b[39m \u001b[38;5;66;03m# Create a set of env runner actors via a EnvRunnerGroup.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m657\u001b[39m \u001b[38;5;28mself\u001b[39m.env_runner_group = \u001b[43mEnvRunnerGroup\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    658\u001b[39m \u001b[43m    \u001b[49m\u001b[43menv_creator\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43menv_creator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    659\u001b[39m \u001b[43m    \u001b[49m\u001b[43mvalidate_env\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mvalidate_env\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    660\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdefault_policy_class\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mget_default_policy_class\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    661\u001b[39m \u001b[43m    \u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    662\u001b[39m \u001b[43m    \u001b[49m\u001b[43mnum_env_runners\u001b[49m\u001b[43m=\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    663\u001b[39m \u001b[43m        \u001b[49m\u001b[32;43m0\u001b[39;49m\n\u001b[32m    664\u001b[39m \u001b[43m        \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    665\u001b[39m \u001b[43m            \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43minput_\u001b[49m\n\u001b[32m    666\u001b[39m \u001b[43m            \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    667\u001b[39m \u001b[43m                \u001b[49m\u001b[38;5;28;43misinstance\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43minput_\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m    668\u001b[39m \u001b[43m                \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    669\u001b[39m \u001b[43m                    \u001b[49m\u001b[38;5;28;43misinstance\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43minput_\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m    670\u001b[39m \u001b[43m                    \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43misinstance\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43minput_\u001b[49m\u001b[43m[\u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m    671\u001b[39m \u001b[43m                \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    672\u001b[39m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    673\u001b[39m \u001b[43m            \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43minput_\u001b[49m\u001b[43m \u001b[49m\u001b[43m!=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43msampler\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\n\u001b[32m    674\u001b[39m \u001b[43m            \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43menable_rl_module_and_learner\u001b[49m\n\u001b[32m    675\u001b[39m \u001b[43m            \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43menable_env_runner_and_connector_v2\u001b[49m\n\u001b[32m    676\u001b[39m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    677\u001b[39m \u001b[43m        \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m.\u001b[49m\u001b[43mnum_env_runners\u001b[49m\n\u001b[32m    678\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    679\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlocal_env_runner\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m    680\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlogdir\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mlogdir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    681\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtune_trial_id\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mtrial_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    682\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    684\u001b[39m \u001b[38;5;66;03m# If an input path is available and we are on the new API stack generate\u001b[39;00m\n\u001b[32m    685\u001b[39m \u001b[38;5;66;03m# an `OfflineData` instance.\u001b[39;00m\n\u001b[32m    686\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[32m    687\u001b[39m     \u001b[38;5;28mself\u001b[39m.config.input_\n\u001b[32m    688\u001b[39m     \u001b[38;5;129;01mand\u001b[39;00m (\n\u001b[32m   (...)\u001b[39m\u001b[32m    697\u001b[39m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m.config.enable_env_runner_and_connector_v2\n\u001b[32m    698\u001b[39m ):\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\rllib\\env\\env_runner_group.py:211\u001b[39m, in \u001b[36mEnvRunnerGroup.__init__\u001b[39m\u001b[34m(self, env_creator, validate_env, default_policy_class, config, num_env_runners, local_env_runner, logdir, _setup, tune_trial_id, num_workers, local_worker)\u001b[39m\n\u001b[32m    198\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m RayActorError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[32m    199\u001b[39m     \u001b[38;5;66;03m# In case of an actor (remote worker) init failure, the remote worker\u001b[39;00m\n\u001b[32m    200\u001b[39m     \u001b[38;5;66;03m# may still exist and will be accessible, however, e.g. calling\u001b[39;00m\n\u001b[32m    201\u001b[39m     \u001b[38;5;66;03m# its `sample.remote()` would result in strange \"property not found\"\u001b[39;00m\n\u001b[32m    202\u001b[39m     \u001b[38;5;66;03m# errors.\u001b[39;00m\n\u001b[32m    203\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m e.actor_init_failed:\n\u001b[32m    204\u001b[39m         \u001b[38;5;66;03m# Raise the original error here that the EnvRunners raised\u001b[39;00m\n\u001b[32m    205\u001b[39m         \u001b[38;5;66;03m# during its construction process. This is to enforce transparency\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m    209\u001b[39m         \u001b[38;5;66;03m# - e.args[0].args[2]: The original Exception (e.g. a ValueError due\u001b[39;00m\n\u001b[32m    210\u001b[39m         \u001b[38;5;66;03m# to a config mismatch) thrown inside the actor.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m211\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m e.args[\u001b[32m0\u001b[39m].args[\u001b[32m2\u001b[39m]\n\u001b[32m    212\u001b[39m     \u001b[38;5;66;03m# In any other case, raise the RayActorError as-is.\u001b[39;00m\n\u001b[32m    213\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    214\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m e\n",
      "\u001b[31mRaySystemError\u001b[39m: System error: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\ntraceback: Traceback (most recent call last):\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 423, in deserialize_objects\n    obj = self._deserialize_object(data, metadata, object_ref)\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 280, in _deserialize_object\n    return self._deserialize_msgpack_data(data, metadata_fields)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 235, in _deserialize_msgpack_data\n    python_objects = self._deserialize_pickle5_data(pickle5_data)\n                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\ray\\_private\\serialization.py\", line 225, in _deserialize_pickle5_data\n    obj = pickle.loads(in_band)\n          ^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\__init__.py\", line 3, in <module>\n    from horcrux_terrain_v1.envs.plane_v1 import PlaneWorld\n  File \"c:\\users\\doore\\documents\\project\\snake_rl\\gd_tor\\learning\\horcrux_terrain_v1\\horcrux_terrain_v1\\envs\\plane_v1.py\", line 2, in <module>\n    from gymnasium.envs.mujoco.mujoco_env import DEFAULT_SIZE\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\__init__.py\", line 1, in <module>\n    from gymnasium.envs.mujoco.mujoco_env import MujocoEnv, MuJocoPyEnv  # isort:skip\n    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\gymnasium\\envs\\mujoco\\mujoco_env.py\", line 19, in <module>\n    import mujoco\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 190, in <module>\n    _load_all_bundled_plugins()\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\site-packages\\mujoco\\__init__.py\", line 183, in _load_all_bundled_plugins\n    PLUGIN_HANDLES.append(ctypes.CDLL(os.path.join(directory, filename)))\n                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"c:\\Users\\doore\\anaconda3\\envs\\gdtor2\\Lib\\ctypes\\__init__.py\", line 376, in __init__\n    self._handle = _dlopen(self._name, mode)\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^\nOSError: [WinError 1114] DLL 초기화 루틴을 실행할 수 없습니다\n"
     ]
    }
   ],
   "source": [
    "config = SACConfig()\n",
    "\n",
    "# 구형 API 구조 사용\n",
    "config.api_stack(\n",
    "    enable_rl_module_and_learner=False,\n",
    "    enable_env_runner_and_connector_v2=False,\n",
    ")\n",
    "\n",
    "config.environment(\"joy-v1\")\n",
    "config.framework(\"torch\")\n",
    "\n",
    "# 병렬 CPU 사용 설정\n",
    "total_workers = 16\n",
    "# config.resources(num_gpus=1,num_cpus_per_worker=1, num_gpus_per_worker= 1/(total_workers+1))\n",
    "config.env_runners(num_env_runners = total_workers, num_cpus_per_env_runner = 1, num_gpus_per_env_runner = 0.3/(total_workers))\n",
    "config.training(\n",
    "    gamma=0.95,\n",
    "    replay_buffer_config={\n",
    "    \"_enable_replay_buffer_api\": True,\n",
    "    \"capacity\": int(4e7),\n",
    "    # If True prioritized replay buffer will be used.\n",
    "    \"prioritized_replay\": False,\n",
    "    \"prioritized_replay_alpha\": 0.6,\n",
    "    \"prioritized_replay_beta\": 0.4,\n",
    "    \"prioritized_replay_eps\": 1e-6,\n",
    "    # Whether to compute priorities already on the remote worker side.\n",
    "    \"worker_side_prioritization\": False,\n",
    "    },\n",
    "\n",
    "    q_model_config = {\n",
    "            \"fcnet_hiddens\": [512, 512, 512, 512, 512, 32],\n",
    "            \"fcnet_activation\": \"tanh\",\n",
    "            \"post_fcnet_hiddens\": [],\n",
    "            \"post_fcnet_activation\": None,\n",
    "            \"custom_model\": None,  # Use this to define custom Q-model(s).\n",
    "            \"custom_model_config\": {},\n",
    "    },\n",
    "    policy_model_config = {\n",
    "            \"fcnet_hiddens\": [512, 512, 512, 512, 512, 32],\n",
    "            \"fcnet_activation\": \"tanh\",\n",
    "            \"post_fcnet_hiddens\": [],\n",
    "            \"post_fcnet_activation\": None,\n",
    "            \"custom_model\": None,  # Use this to define a custom policy model.\n",
    "            \"custom_model_config\": {},\n",
    "    },\n",
    "\n",
    "    train_batch_size_per_learner = 4096,\n",
    "    num_steps_sampled_before_learning_starts = 4096,\n",
    ")\n",
    "\n",
    "algo = config.build()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11ae756f-bb08-4813-bf07-c019392dac99",
   "metadata": {},
   "source": [
    "## 학습시작"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2ae0dfe-e461-442b-b090-cc437723543f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pprint import pprint\n",
    "\n",
    "n_iter = 63000\n",
    "save_iter = 0\n",
    "save_name = \"SAC_layer_512_5_32_linear_friction_joy_31\"\n",
    "\n",
    "for i in range(n_iter):\n",
    "    result = algo.train()\n",
    "    print(f\"{i:03d}\", end=\", \")\n",
    "    # result.pop(\"config\")\n",
    "    # pprint(result)\n",
    "\n",
    "    if i%3000 == 0:\n",
    "        checkpoint_dir = algo.save(save_name+\"_\"+str(save_iter))\n",
    "        print(f\"Checkpoint saved in directory {checkpoint_dir}\")\n",
    "        save_iter += 1\n",
    "\n",
    "algo.save(save_name+str(\"_final\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc4f40a7-044a-4e2c-a8c0-fcb8575b5bde",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "# 학습 알고리즘 평가하기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dcfb0e4-1a84-492f-a0d7-9787b0182bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "algo = Algorithm.from_checkpoint(\"./SAC_layer_512_5_32_linear_friction_joy2_2\") # 학습된 정책 로드"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27e963f9-0653-4c35-a7f4-3fb03d534bfe",
   "metadata": {},
   "source": [
    "## 평가용 Env 생성 및 실행"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d825c204-3cca-4c0d-878c-a68833f12909",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import torch\n",
    "import numpy as np\n",
    "import gymnasium as gym\n",
    "from horcrux_terrain_v1.envs import SandWorld\n",
    "from ray.rllib.algorithms.algorithm import Algorithm\n",
    "import time\n",
    "\n",
    "eval_config = env_config\n",
    "eval_config[\"use_friction_chg\"] = False\n",
    "eval_config[\"joy_input_random\"] = False\n",
    "eval_config[\"joy_input\"] = (1, 0, 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db4b174b-71a4-4a63-8788-02fb73512db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "env = gym.make(\"horcrux_terrain_v1/plane-v2\", \n",
    "               terminate_when_unhealthy = False, \n",
    "               render_mode = \"human\", \n",
    "               render_camera_name = 'ceiling', \n",
    "               use_gait = True,               \n",
    "               **eval_config,\n",
    "               ) \n",
    "\n",
    "for j in range(3):\n",
    "   episode_return = 0\n",
    "   yaw_total = 0\n",
    "   terminated = truncated = False\n",
    "\n",
    "   obs, info = env.reset()\n",
    "\n",
    "   for i in range(1000):\n",
    "\n",
    "      action = algo.compute_single_action(observation= obs)\n",
    "      \n",
    "      obs, reward, terminated, truncated, info = env.step(action)\n",
    "      if i == 0:\n",
    "          print(f\"Joy input = {info['joy_input']}\")\n",
    "      if terminated:\n",
    "         print(\"terminated\")\n",
    "\n",
    "      episode_return += reward\n",
    "      yaw_total += info['step_ypr'][0] / 0.1\n",
    "\n",
    "   print(f\"Reached episode return of {episode_return}.\")\n",
    "   print(yaw_total)\n",
    "\n",
    "env.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd1aaeb7-4697-41bc-ad9b-079f609dc5ed",
   "metadata": {},
   "source": [
    "### Pytorch 모델 바로 불러오기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "087f2ca1-c6d4-477d-a88b-9ee23e296869",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "from ray.rllib.policy.policy import Policy\n",
    "from ray.rllib.models.torch.torch_modelv2 import TorchModelV2\n",
    "\n",
    "_algo = Algorithm.from_checkpoint(\"./Paper_agents/good/Linear/SAC_layer_512_5_32_Linear_restart_final\")\n",
    "# _policy = Policy.from_checkpoint(\"./SAC_layer_512_5_32_linear_friction_joy_31_0\")\n",
    "# _policy = Policy.from_checkpoint(\"./Paper_agents/good/Linear/SAC_layer_512_5_32_Linear_restart_final\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d7d3aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "randobs = np.random.random(94)\n",
    "\n",
    "# _algo.compute_single_action(randobs)\n",
    "# _algo.get_policy('default_policy').dist_class.mro()\n",
    "_algo.get_policy('default_policy').model.action_dist\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5545bfe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "randobs = np.random.random(94)\n",
    "\n",
    "# obs = torch.tensor(randobs, dtype=torch.float32)\n",
    "_policy['default_policy'].dist_class\n",
    "# _policy['default_policy'].compute_single_action(torch.tensor(randobs, dtype=torch.float32), explore=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02b1a761",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gdtor2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
